File: AssetManager.Api/Attributes/AssetNameAttribute.cs
//
//          Copyright Seth Hendrick 2019.
// Distributed under the Boost Software License, Version 1.0.
//    (See accompanying file LICENSE_1_0.txt or copy at
//          http://www.boost.org/LICENSE_1_0.txt)
//

namespace AssetManager.Api.Attributes
{
    public class AssetNameAttribute : IAttribute
    {
        // ---------------- Constructor ----------------

        public AssetNameAttribute() :
            this( string.Empty )
        {
        }

        public AssetNameAttribute( string assetName )
        {
            this.Value = assetName;
        }

        // ---------------- Properties ----------------

        AttributeTypes IAttribute.AttributeType
        {
            get
            {
                return AttributeTypes.AssetName;
            }
        }

        /// <summary>
        /// The name of the asset instance.
        /// </summary>
        public string Value { get; set; }

        // ---------------- Functions ----------------

        public override string ToString()
        {
            return this.Value.ToString();
        }

        void IAttribute.Deserialize( string serialData )
        {
            this.Value = serialData ;
        }

        string IAttribute.Serialize()
        {
            return this.Value;
        }
    }
}

File: AssetManager.Web/Scripts/AssetAdder/AssetNameAttribute.ts
//
//          Copyright Seth Hendrick 2019.
// Distributed under the Boost Software License, Version 1.0.
//    (See accompanying file LICENSE_1_0.txt or copy at
//          http://www.boost.org/LICENSE_1_0.txt)
//

class AssetNameAttribute extends BaseAttribute {

    // ---------------- Fields ----------------

    private value: string;

    // ---------------- Constructor ----------------

    constructor(info: AssetNameAttributeTypeInfo, key: string) {
        super(info.AttributeType, key)
    }

    // ---------------- Functions ----------------

    public Validate(): Array<string> {
        let errors = new Array<string>();

        if (Helpers.StringIsNullOrEmpty(this.value)) {
            errors.push("No value specified.");
        }

        return errors;
    }

    public ToJson(): object {
        let data = {
            "Key": this.Key,
            "AttributeType": this.AttributeType,
            "Value": this.GetValue()
        };

        return data;
    }

    // -------- Setters --------

    public SetValue(newValue: string): AssetNameAttribute {
        if (newValue === "") {
            this.value = null;
        }
        else {
            this.value = newValue;
        }
        return this;
    }

    // -------- Getters --------

    public GetValue(): string {
        return this.value;
    }
}